### ISSUE TITLE
Conch blows up when processes end due to signals

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#2687]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)| @jml reported|
|-|-|
|Trac ID|trac#2687|
|Type|defect|
|Created|2007-06-01 04:17:16Z|
|Branch|https://github.com/twisted/twisted/tree/session-signal-2687|

[RFC 4254](http://www.ietf.org/rfc/rfc4254.txt), Section 6.10 says:

```
   The remote command may also terminate violently due to a signal.
   Such a condition can be indicated by the following message.  A zero
   'exit_status' usually means that the command terminated successfully.

      byte      SSH_MSG_CHANNEL_REQUEST
      uint32    recipient channel
      string    "exit-signal"
      boolean   FALSE
      string    signal name (without the "SIG" prefix)
      boolean   core dumped
      string    error message in ISO-10646 UTF-8 encoding
      string    language tag [RFC3066]
```

`SSHSessionProcessProtocol` says:
```python
    def processEnded(self, reason = None):
        if reason and hasattr(reason.value, 'exitCode'):
            log.msg('exitCode: %s' % repr(reason.value.exitCode))
            self.session.conn.sendRequest(self.session, 'exit-status', struct.pack('!L', reason.value.exitCode))
        self.session.loseConnection()
```

Note that the `ProcessTerminated` exception will always have an `exitCode` attribute. If the process terminated due to a signal, then the `exitCode` attribute will be `None`, which cannot be packed into the struct. Conch should send an `exit-signal` message instead of an `exit-status` messge when `exitCode` is `None`.

<details><summary>Searchable metadata</summary>

```
trac-id__2687 2687
type__defect defect
reporter__jml jml
priority__normal normal
milestone__ 
branch__branches_session_signal_2687 branches/session-signal-2687
branch_author__therve therve
status__closed closed
resolution__fixed fixed
component__conch conch
keywords__ 
time__1180671436000000 1180671436000000
changetime__1246456871000000 1246456871000000
version__None None
owner__ 

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_1"></a>|@jml commented|
|-|-|

Basic patch to solve the explosion.

```
=== modified file 'twisted/conch/ssh/session.py'
--- twisted/conch/ssh/session.py        2006-03-16 08:44:37 +0000
+++ twisted/conch/ssh/session.py        2007-06-04 07:36:55 +0000
@@ -193,10 +193,14 @@
     def connectionLost(self, reason = None):
         self.session.loseConnection()
 
-    def processEnded(self, reason = None):
-        if reason and hasattr(reason.value, 'exitCode'): 
-            log.msg('exitCode: %s' % repr(reason.value.exitCode))
-            self.session.conn.sendRequest(self.session, 'exit-status', struct.pack('!L', reason.value.exitCode))
+    def processEnded(self, reason=None):
+        if reason:
+            exitCode = getattr(reason.value, 'exitCode', None)
+            if exitCode is not None:
+                log.msg('exitCode: %s' % repr(reason.value.exitCode))
+                self.session.conn.sendRequest(
+                    self.session, 'exit-status',
+                    struct.pack('!L', reason.value.exitCode))
         self.session.loseConnection()
 
     # transport stuff (we are also a transport!)
```

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_3"></a>|@therve commented|
|-|-|

I don't see any test.

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_4"></a>|@jml commented|
|-|-|

Right. I'd appreciate help writing the test.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_5"></a>|@therve set owner to @therve|
|-|-|

I guess I'll try.

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_6"></a>|@jml commented|
|-|-|

You're a champion :)

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_7"></a>|@therve commented|
|-|-|

This is done correctly (almost) in [#2710](https://github.com/twisted/twisted/issues/2710). I think all the efforts should be into that direction.

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_8"></a>|@jml commented|
|-|-|

This is a pretty trivial bug to fix, has been open for a year and is still biting us. I hereby express my renewed interest in fixing it personally.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_10"></a>|@therve commented|
|-|-|

(In [27051]) Branching to 'session-signal-2687'

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_11"></a>|@therve removed owner|
|-|-|

Let's revive that, from [#2710](https://github.com/twisted/twisted/issues/2710) grave.

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_12"></a>|@jml commented|
|-|-|

I did a review of the branch face-to-face with Thomas, and he's corrected all the comments I've had.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_14"></a>|@therve set status to `closed`|
|-|-|

(In [27060]) Merge session-signal-2687

Author: therve
Reviewer: jml
Fixes [#2687](https://github.com/twisted/twisted/issues/2687)

Manage process termination in Conch due to signals, sending exit-signal
message instead of exit-status.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
